drivers/flash/flash_andes_qspi_xip.c

```diff
@@ -0,0 +1,499 @@
+		ret = write_protection_set(dev, false);
+		if (ret != 0) {
+			break;
+		}
+
+		/* Get the adequate size to send*/
+		to_write = MIN(PAGE_SIZE - (addr_curr % PAGE_SIZE), size_remainig);
+		ret = flash_andes_qspi_xip_cmd_addr_write(dev, SPI_NOR_CMD_PP_1_1_4, addr_curr, src,
+							  to_write);
+		flash_andes_qspi_xip_wait_until_ready(dev);
+
+		if (ret != 0) {
+			break;
+		}
+
+		src = (const uint8_t *)src + to_write;
+		size_remainig -= to_write;
+		addr_curr += to_write;
+	}
+
+	/* Make sure write protection is enabled at the end of write. */
+	int ret2 = write_protection_set(dev, true);
+
+	if (!ret) {
+		ret = ret2;
+	}
+
+	cleanup_after_flashing(dev, addr, size);
+
+	return ret;
+}
+
+static int flash_andes_qspi_xip_write(const struct device *dev, off_t addr, const void *src,
+				      size_t size)
+{
+	const struct flash_andes_qspi_xip_config *config = dev->config;
+	int ret = 0;
+	unsigned int key;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	/* Check write boundaries */
+	if ((addr < 0 || addr >= config->flash_size) || ((config->flash_size - addr) < size)) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Synchronous mechanisms like semaphores are not needed,
+	 * because interrupts are always locked and there are no reschedule points.
+	 */
+	key = irq_lock();
+
+	ret = do_write(dev, addr, src, size);
+
+	irq_unlock(key);
+
+	return ret;
+}
+
+static __ramfunc int do_erase(const struct device *dev, off_t addr, size_t size)
+{
+	int ret;
+	off_t addr_curr = addr;
+	size_t size_remainig = size;
+
+	prepare_for_flashing();
+
+	while (size_remainig > 0) {
+		ret = write_protection_set(dev, false);
+		if (ret != 0) {
+			break;
+		}
+
+		/* Use the smallest erase unit not to hold CPU for long. */
+		ret = flash_andes_qspi_xip_cmd_addr_write(dev, SPI_NOR_CMD_SE, addr_curr, NULL, 0);
+		flash_andes_qspi_xip_wait_until_ready(dev);
+		if (ret != 0) {
+			break;
+		}
+		addr_curr += SPI_NOR_SECTOR_SIZE;
+		size_remainig -= SPI_NOR_SECTOR_SIZE;
+	}
+
+	/* Make sure write protection is enabled at the end of erase. */
+	int ret2 = write_protection_set(dev, true);
+
+	if (!ret) {
+		ret = ret2;
+	}
+
+	cleanup_after_flashing(dev, addr, size);
+
+	return ret;
+}
+
+static int flash_andes_qspi_xip_erase(const struct device *dev, off_t addr, size_t size)
+{
+	const struct flash_andes_qspi_xip_config *config = dev->config;
+	int ret = 0;
+	unsigned int key;
+
+	if (size == 0) {
+		return 0;
+	}
+
+	/* Check erase boundaries */
+	if ((addr < 0 || addr >= config->flash_size) || ((config->flash_size - addr) < size)) {
+		return -EINVAL;
+	}
+
+	/* address must be sector-aligned */
+	if (!SPI_NOR_IS_SECTOR_ALIGNED(addr)) {
+		return -EINVAL;
+	}
+
+	/* size must be a multiple of sectors */
+	if ((size % SPI_NOR_SECTOR_SIZE) != 0) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Synchronous mechanisms like semaphores are not needed,
+	 * because interrupts are always locked and there are no reschedule points.
+	 */
+	key = irq_lock();
+
+	do_erase(dev, addr, size);
```

Comment:
Similarly, it is recommended to disable mmisc_ctl.BRPE on the Andes platform before jumping to __ramfunc, and re-enable it after returning.